src/antlr/index.ts

```diff
@@ -15,20 +15,32 @@ interface ParserOptions {
   cursorPosition?: { lineNumber: number; column: number };
 }
 
-type TableProvider<T, C> = (
-  id: TablePrimary
-) => { table: { id: string; data: T }; columns: { id: string; data: C }[] } | undefined;
+interface Named {
+  name: string;
+}
+
+export interface MetadataProvider<
+  Catalog extends Named = Named,
+  Schema extends Named = Named,
+  Table extends Named = Named,
+  Column extends Named = Named
+> {
+  getCatalogs: () => Catalog[] | undefined;
+  getSchemas: (arg?: { catalog: string }) => Schema[] | undefined;
+  getTables: (args?: { catalogOrSchema: string; schema?: string }) => Table[] | undefined;
+  getColumns: (args: { table: string; catalogOrSchema?: string; schema?: string }) => Column[] | undefined;
+}
```

Comment:
@megaflo Can you also work with an empty array here? Usually we try to avoid `[] | undefined'.

Reply:
For `getColumns` we introduced the capabilities to deduce column names if fetching has somehow failed. For this we need to distinguish between an empty table and a failed fetch job.

I don't mind removing the undefined from the return type, but it will restrict us slightly when it comes to deriving information from the query.

Reply:
@megaflo interesting. How can fetching fail? Hub backend down - that kind of issue?

How do we deduce column names?

Reply:
Yeah, either a technical failure or even just a misspelled table. In https://github.com/contiamo/rhombic/pull/38 we introduced the ability to infer columns in such a case. This way we can still show something useful in the lineage. The inference is only working locally right now (so only within the subquery where the offending table is used) so it doesn't really add much for auto completion. Well, we could then suggest the same column names in different places in the same query, but that's arguably not the most important feature.